### ISSUE TITLE
Generic subclass of a generic class, using a constrained TypeVar is mishandled

### ISSUE THREAD

**Issue by ANogin** [python/mypy#17757]
**Bug Report**

``` python
from typing import Generic, TypeVar

T = TypeVar("T")
N = TypeVar("N", int, float)

class C(Generic[T]):
    def __init__(self, c: T):
        self.c = c

class C2(C[N]):
    def __init__(self, c: N):
        super().__init__(c) # Unexpected error
```

results in an unexpected
```
error: Argument 1 to "__init__" of "C" has incompatible type "int"; expected "N"  [arg-type]
error: Argument 1 to "__init__" of "C" has incompatible type "float"; expected "N"  [arg-type]
```
on the last line of the test file.

**Your Environment**

Confirmed with mypy 1.11.2 in Python 3.11.9, without any flags; does not seem version-specific.

**Comment by ANogin** [comment]
Still the same behavior with Python 3.13.5 mypy 1.17.0
